### PR TITLE
Turn illegal rewrite rules into an error warning

### LINKED ISSUE #6643: Rewrite rules are allowed in implicit mutual blocks
```agda
{-# OPTIONS --rewriting --confluence-check #-}
module Test45 where

open import Agda.Builtin.Equality
open import Agda.Builtin.Bool

{-# BUILTIN REWRITE _≡_ #-}

p : true ≡ false
q : true ≡ false
p = q
{-# REWRITE p #-}

q = refl
```

### LINKED ISSUE #7090: REWRITE rule with confluence, inconsistencies with documentation and error messages
I was following the documentation on rewriting, specifically [confluence checking](https://agda.readthedocs.io/en/v2.6.4.1/language/rewriting.html#confluence-checking), which states

> [...] For example, if there are two rules suc m + n = suc (m + n) and m + suc n = suc (m + n), then there should also be a rule suc m + suc n = suc (suc (m + n)).

I tried this via the following code (note `--local-confluence-check`):

```agda
{-# OPTIONS --rewriting --local-confluence-check  #-}

module RewriteTest0 where

open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Data.Nat
open import Data.Nat.Properties
open import Relation.Binary.PropositionalEquality

+-suc-suc : ∀ (m n : ℕ) → suc m + suc n ≡ suc (suc (m + n))
+-suc-suc m n rewrite +-suc m n = refl

{-# REWRITE +-identityʳ +-suc +-suc-suc #-}

```

This gives me the following error message:

```
+-suc-suc  is not a legal rewrite rule, since the left-hand side 
suc m + suc n  reduces to  suc (m + suc n)
when checking the pragma REWRITE +-identityʳ +-suc +-suc-suc
```

I think this implies a bug or that the documentation is wrong or misleading. Any feedback would be appreciated.

The following works fine:

```agda
{-# OPTIONS --rewriting --local-confluence-check #-}

module RewriteTest1 where

open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Data.Nat
open import Data.Nat.Properties
open import Relation.Binary.PropositionalEquality

{-# REWRITE +-identityʳ +-suc #-}

+comm : ∀ {m n : ℕ} → m + n ≡ n + m
+comm {zero} = refl
+comm {suc m} = cong suc (+comm {m})
```

When changing `--local-confluence-check` to `--confluence-check` in `RewriteTest1` , I get the following error messages:

```
Global confluence check failed: suc n + suc n₁ can be rewritten to
either suc (suc n + n₁) or suc (n + suc n₁).
Possible fix: add a rewrite rule with left-hand side suc n + suc n₁
to resolve the ambiguity.
when checking confluence of the rewrite rule +-suc with
Agda.Builtin.Nat._+_-clause2

Global confluence check failed: zero + suc n can be rewritten to
either suc (zero + n) or suc n.
Possible fix: add a rewrite rule with left-hand side zero + suc n
to resolve the ambiguity.
when checking confluence of the rewrite rule +-suc with
Agda.Builtin.Nat._+_-clause1

Global confluence check failed: suc n + 0 can be rewritten to
either suc n or suc (n + 0).
Possible fix: add a rewrite rule with left-hand side suc n + 0 to
resolve the ambiguity.
when checking confluence of the rewrite rule +-identityʳ with
Agda.Builtin.Nat._+_-clause2
```

Changing my code to include the equalities suggested by the error messages makes Agda accept the program. (I am confused, however, as two of them seem to just be special cases of existing rules, see code comments below. Also, the documentation does not mention these being needed.)

```agda
{-# OPTIONS --rewriting --confluence-check #-}

module RewriteTest2 where

open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Data.Nat
open import Data.Nat.Properties
open import Relation.Binary.PropositionalEquality

-- the documentation mentioned, that this rule was needed
+-suc-suc : ∀ (m n : ℕ) → suc m + suc n ≡ suc (suc (m + n))
+-suc-suc m n rewrite +-suc m n = refl

-- we already know this from the definition of ℕ, not sure why we need this
zero+suc-n : ∀ (n : ℕ) → 0 + suc n ≡ suc n
zero+suc-n n = refl

-- a specialisation of identityʳ, not sure why we need this
suc-n+zero : ∀ (n : ℕ) → suc n + 0 ≡ suc n
suc-n+zero n rewrite +-identityʳ n = refl

{-# REWRITE +-identityʳ +-suc +-suc-suc zero+suc-n suc-n+zero #-}

+comm : ∀ {m n : ℕ} → m + n ≡ n + m
+comm {zero} = refl
+comm {suc m} = cong suc (+comm {m})

```

Changing the flag in `RewriteTest2` to `--local-confluence-check`, however, gives an error message:

```
+-suc-suc  is not a legal rewrite rule, since the left-hand side 
suc m + suc n  reduces to  suc (m + suc n)
when checking the pragma
REWRITE +-identityʳ +-suc +-suc-suc zero+suc-n suc-n+zero
```
This, despite the documentation saying that `--local-confluence-check` is less restrictive.

##### Addenda

This came from me trying to minimise an error I was getting while using `REWRITE` in a project, which unfortunately I cannot share at this moment. In that project, enabling `--confluence-check` gives the following error:
```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Rewriting/Confluence.hs:606:19 in Agd-2.6.4.1-d9da44fe:Agda.TypeChecking.Rewriting.Confluence
```
When using holes in a theorem that uses the rewrite rules, I get the following error:
```
Panic: uncaught pattern violation
```

```
$ agda --version     
Agda version 2.6.4.1
```

### PR DESCRIPTION
**Pull request by jespercockx**
This PR turns all errors that can be thrown during the checking of a rewrite rule into warnings. If such a warning is thrown, the rewrite rule is simply not added, but the rest of the file can still be checked.

It is based on #7341, and supersedes #7342.

### CODE REVIEW

File: doc/user-manual/tools/command-line-options.rst [hunk 1]
```diff
@@ -1491,6 +1491,66 @@ The list containing any warning ``NAME`` can be produced by ``agda --help=warnin
 
      :ref:`NO_TERMINATION_CHECK<terminating-pragma>` pragmas; such are deprecated.
 
+.. option:: RewriteLHSNotDefOrConstr
+
+     Rewrite rule head symbol is not a defined symbol or constructor.
```
Comment (andreasabel) [inline doc/user-manual/tools/command-line-options.rst:1496]: From the explanation, this looks like a duplicate of `RewriteHeadSymbolNotPostulateFunctionConstructor`.
Also, maybe `Def` and `Constr` should be expanded (unabbreviated) for sake of uniformity.

File: doc/user-manual/tools/command-line-options.rst [hunk 2]
```diff
@@ -1491,6 +1491,66 @@ The list containing any warning ``NAME`` can be produced by ``agda --help=warnin
...
+
+.. option:: RewriteHeadSymbolIsProjectionLikeFunction
+
+     Rewrite rule head symbol is a projection-like function.
+
+.. option:: RewriteHeadSymbolNotPostulateFunctionConstructor
+
+     Rewrite rule head symbol is not a postulate, function, or constructor.
+
+.. option:: RewriteHeadSymbolDefContainsMetas
```
Comment (andreasabel) [inline doc/user-manual/tools/command-line-options.rst]: Drop `Def` (or expand to `Definition`)?

File: doc/user-manual/tools/command-line-options.rst [hunk 3]
```diff
@@ -1491,6 +1491,66 @@ The list containing any warning ``NAME`` can be produced by ``agda --help=warnin
...
+
+     Definition of rewrite rule head symbol contains unsolved metas.
+
+.. option:: RewriteConstructorParamsNotGeneral
```
Comment (andreasabel) [inline doc/user-manual/tools/command-line-options.rst]: Expand `Params` to `Parameters`?

**Review by andreasabel — COMMENTED** [review]
Thanks, great stuff!

Approved modulo some bikeshedding of warning names.

File: doc/user-manual/tools/command-line-options.rst (hunk 1 above)
Comment (jespercockx) [reply doc/user-manual/tools/command-line-options.rst:1496]: Actually the other error (`RewriteHeadSymbolNotPostulateFunctionConstructor`) is used only when the head symbol of a rewrite rule is a type constructor, so I'll rename that one to `RewriteHeadSymbolIsTypeConstructor` and expand the Def and Constr in the name of this one.

**Comment by jespercockx** [comment]
@andreasabel I addressed your comments! Should this be rebased or squashed?

**Comment by jespercockx** [comment]
Perhaps the best solution is to first merge #7341 and then this one, squashing each PR so there are two commits on master: one adding the new error, and the second one turning all errors into warnings.

**Review by andreasabel — APPROVED** [review]

**Pull request merged**